### PR TITLE
[ioctl] build xrc20 transfer from command line into new ioctl

### LINKED ISSUE #3339: [ioctl] build xrc20 transfer from command line into new ioctl
### What would you like to be added
 - Step 1: create `xrc20transferfrom.go` and `xrc20transferfrom_test.go` files in `./ioctl/newcmd/action`. create this directory if not exist. 
 - Step 2: refactor  command  https://github.com/iotexproject/iotex-core/blob/master/ioctl/cmd/action/xrc20transferfrom.go in `xrc20transferfrom.go`
 - Step 3: add unit test in xx_test.go

#### Notes:
1. Use [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) to construct the Cobra command.
2. [Output package](https://github.com/iotexproject/iotex-core/blob/master/ioctl/output) is deprecated. `e.g.` `output.PrintError` could be replaced with `errors.Wrap`.
3. Global variables (config.ReadConfig  config.DefaultConfigFile etc.) have been wrapped into client interface. Please use [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) to access them if needed
4. The logic of new code should be similar to [the old one](https://github.com/iotexproject/iotex-core/blob/master/ioctl/cmd/action/xrc20transferfrom.go).
5. In each unit test, the results should be checked whether they are as the expected. 

#### Example:
1. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountinfo.go
2. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountdelete.go
3. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountcreateadd.go

### Why is this needed
Unit tests couldn't be added for the old ioctl command (./ioctl/cmd)

### PR DESCRIPTION
**Pull request by saitofun**
# Description
[ioctl] build xrc20 transfer from command line into new ioctl

Fixes #3339 

## Type of change
Please delete options that are not relevant.
- [x] Code refactor or improvement

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: ioctl/newcmd/action/action.go [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+	if util.AliasIsHdwalletKey(addressOrAlias) {
+		return addressOrAlias, nil
+	}
+
+	if addressOrAlias == "" {
+		addressOrAlias, err = config.GetContextAddressOrAlias()
+		if err != nil {
+			return
+		}
+	}
```
Comment (huof6829) [inline ioctl/newcmd/action/action.go:253]: these are done in line 254,

File: ioctl/newcmd/action/xrc20.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strconv"
+
+	"github.com/iotexproject/iotex-core/ioctl"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	"github.com/iotexproject/iotex-address/address"
```
Comment (huof6829) [inline ioctl/newcmd/action/xrc20.go:19]: these are another single group

File: ioctl/newcmd/action/xrc20.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+	"github.com/iotexproject/iotex-core/ioctl/cmd/alias"
```
Comment (huof6829) [inline ioctl/newcmd/action/xrc20.go:21]: don't use  `iotex-core/ioctl/cmd/`,  use `iotex-core/ioctl/newcmd/`

File: ioctl/newcmd/action/xrc20const.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+				"name": "value",
+				"type": "uint256"
+			}
+		],
+		"name": "Transfer",
+		"type": "event"
+	}
+]`
+
+var _xrc20ABI abi.ABI
```
Comment (huof6829) [inline ioctl/newcmd/action/xrc20const.go:187]: put global variable into client.

File: ioctl/newcmd/action/xrc20transferfrom.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+			}
+			contract, err := xrc20Contract(cmd)
+			if err != nil {
+				return errors.Wrap(err, "failed to get contract address")
+			}
+			amount, err := parseAmount(client, cmd, contract, args[2])
+			if err != nil {
+				return errors.Wrap(err, "failed to parse amount")
+			}
+			bytecode, err := _xrc20ABI.Pack("transferFrom", owner, recipient, amount)
```
Comment (huof6829) [inline ioctl/newcmd/action/xrc20transferfrom.go:59]: put line59 into client and use like client.PackABI

**Comment by huof6829** [comment]
create _test.go and add unit test.

File: ioctl/newcmd/action/action.go [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+	return nil
+}
+
+type sendMessage struct {
+	Info   string `json:"info"`
+	TxHash string `json:"txHash"`
+	URL    string `json:"url"`
+}
+
+func (m *sendMessage) String() string {
```
[the remark this answers is not in the document]
Comment (saitofun) [reply ioctl/newcmd/action/action.go:525]: why make `PackABI` as a client member method?

File: ioctl/client_abi.go [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
...},{"name":"value","type":"uint256"}],"name":"transferFrom","outputs":[{"name":"","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"who","type":"address"}],"name":"balanceOf","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"to","type":"address"},{"name":"value","type":"uint256"}],"name":"transfer","outputs":[{"name":"","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"owner","type":"address"},{"name":"spender","type":"address"}],"name":"allowance","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"anonymous":false,"inputs":[{"indexed":true,"name":"owner","type":"address"},{"indexed":true,"name":"spender","type":"address"},{"indexed":false,"name":"value","type":"uint256"}],"name":"Approval","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"from","type":"address"},{"indexed":true,"name":"to","type":"address"},{"indexed":false,"name":"value","type":"uint256"}],"name":"Transfer","type":"event"}]`
```
Comment (huof6829) [inline ioctl/client_abi.go:12]: not global.  should be client's member.

**Comment by huof6829** [comment]
@saitofun   fix ci Fail

**Comment by huof6829** [comment]
@saitofun   I fixed,  pls check it. Thx

**Comment by huof6829** [comment]
@saitofun   action_test.go is handling by  https://github.com/iotexproject/iotex-core/pull/3472. you can refer to it.

**Comment by huof6829** [comment]
xrc20transferfrom_test.go:98   run failed

**Comment by saitofun** [comment]
> xrc20transferfrom_test.go:98 run failed

already fix it

**Review by huof6829 — APPROVED** [review]

File: ioctl/client.go [hunk 8]
```diff
@@ -122,16 +131,21 @@ func EnableCryptoSm2() Option {
 }
 
 // NewClient creates a new ioctl client
-func NewClient(cfg config.Config, configFilePath string, opts ...Option) Client {
+func NewClient(cfg config.Config, configFilePath string, opts ...Option) (Client, error) {
+	xrc20ABI, err := abi.JSON(strings.NewReader(_xrc20ABI))
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get abi JSON data")
+	}
```
Comment (CoderZhi) [inline ioctl/client.go:138]: we may put it into `init`, and panic if any error

File: ioctl/client.go [hunk 9]
```diff
@@ -397,6 +426,10 @@ func (c *client) WriteHdWalletConfigFile(mnemonic string, password string) error
 	return nil
 }
 
+func (c *client) PackABI(name string, args ...interface{}) ([]byte, error) {
+	return c.xrc20ABI.Pack(name, args...)
```
Comment (CoderZhi) [inline ioctl/client.go:430]: xrc20abi is more like a const, instead of a parameter of client

File: ioctl/newcmd/action/xrc20.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+	cmd.AddCommand(NewXrc20TransferFromCmd(client))
+	// TODO cmd.AddCommand(NewXrc20TotalSupply(client))
+	// TODO cmd.AddCommand(NewXrc20BalanceOf(client))
+	// TODO cmd.AddCommand(NewXrc20Transfer(client))
+	// TODO cmd.AddCommand(NewXrc20Approve(client))
+	// TODO cmd.AddCommand(NewXrc20Allowance(client))
+
+	client.SetEndpointWithFlag(cmd.PersistentFlags().StringVar)
+	client.SetInsecureWithFlag(cmd.PersistentFlags().BoolVar)
+	client.SetXrc20ContractAddrWithFlag(cmd.PersistentFlags().StringVarP, cmd.MarkFlagRequired)
```
Comment (CoderZhi) [inline ioctl/newcmd/action/xrc20.go:50]: xrc20 contract address is not a client feature.

File: ioctl/client.go (hunk 9 above)
Comment (huof6829) [reply ioctl/client.go:455]: `xrc20ABI.Pack(name, args...)`  is used in other xrc20 commands: xrc20_allowance、 xrc20_approve ....
so make this function. 
It seems that  xrc20abi has nothing to do with client. If so, I will put it into package of xrc20.

File: ioctl/newcmd/action/xrc20.go (hunk 10 above)
Comment (huof6829) [reply ioctl/newcmd/action/xrc20.go:50]: ok,

File: ioctl/client.go (hunk 8 above)
Comment (huof6829) [reply ioctl/client.go]: fixed

**Review by Liuhaai — APPROVED** [review]

**Comment by Liuhaai** [comment]
@huof6829 could you help fix the ci failure?

**Pull request merged**